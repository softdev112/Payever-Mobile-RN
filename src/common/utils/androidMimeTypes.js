const mimeTypes = {
  mp3:     'audio/mpeg',
  mpga:    'audio/mpeg',
  m4a:     'audio/mp4',
  wav:     'audio/x-wav',
  amr:     'audio/amr',
  awb:     'audio/amr-wb',
  wma:     'audio/x-ms-wma',
  ogg:     'audio/ogg, application/ogg',
  aac:     'audio/aac, audio/aac-adts',
  mka:     'audio/x-matroska',
  mid:     'audio/midi',
  midi:    'audio/midi',
  xmf:     'audio/midi',
  rtttl:   'audio/midi',
  smf:     'audio/sp-midi',
  imy:     'audio/imelody',
  rtx:     'audio/midi',
  ota:     'audio/midi',
  mxmf:    'audio/midi',
  mpeg:    'video/mpeg, video/mp2p',
  mpg:     'video/mpeg, video/mp2p',
  mp4:     'video/mp4',
  m4v:     'video/mp4',
  '3gp':   'video/3gpp',
  '3gpp':  'video/3gpp',
  '3G2':   'video/3gpp2',
  '3GPP2': 'video/3gpp2',
  mkv:     'video/x-matroska',
  webm:    'video/webm',
  ts:      'video/mp2ts',
  avi:     'video/avi',
  wmv:     'video/x-ms-wmv',
  asf:     'video/x-ms-asf',
  jpg:     'image/jpeg',
  jpeg:    'image/jpeg',
  gif:     'image/gif',
  png:     'image/png',
  bmp:     'image/x-ms-bmp',
  wbmp:    'image/vnd.wap.wbmp',
  webp:    'image/webp',
  m3u:     'audio/x-mpegurl, application/x-mpegurl',
  pls:     'audio/x-scpls',
  wpl:     'application/vnd.ms-wpl',
  m3u8:    'application/vnd.apple.mpegurl, audio/mpegurl, audio/x-mpegurl',
  fl:      'application/x-android-drm-fl',
  txt:     'text/plain',
  htm:     'text/html',
  html:    'text/html',
  pdf:     'application/pdf',
  doc:     'application/msword',
  xls:     'application/vnd.ms-excel',
  ppt:     'application/mspowerpoint',
  flac:    'audio/flac',
  zip:     'application/zip',
  rar:     'application/zip',
};

function getTypeByExt(ext: string) {
  if (!ext || ext === '') return '*/*';

  let fileExt = ext;
  if (ext.startsWith('.')) {
    fileExt = ext.substr(1).toLowerCase();
  }

  return mimeTypes[fileExt] ? mimeTypes[fileExt] : '*/*';
}

export default {
  getTypeByExt,
};