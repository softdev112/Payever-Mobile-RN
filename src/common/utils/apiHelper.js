import { runInAction } from 'mobx';
import * as log from './log';

/**
 * This helper simplifies API call from Store
 */
export class ApiHelper {
  apiPromise: Promise<ApiResp>;
  fetchPromise: Promise;
  returnValue: any;
  store: ?Object = null;

  onSuccess: (resp: ApiResp) => any;
  onError: (e: ErrorObject) => any;
  onComplete: () => any;

  constructor(apiPromise: Promise<ApiResp>, store: Object = null) {
    this.apiPromise = apiPromise;
    this.store = store;

    this.onSuccess  = () => {};
    this.onError    = () => {};
    this.onComplete = () => {};
  }

  /**
   * Call this callback if API promise has been resolved
   * Returned value is saved and can be returned in a promise through promise()
   */
  success(callback: (resp: ApiResp) => any): ApiHelper {
    this.onSuccess = callback;
    return this;
  }

  /**
   * Call this callback if API promise has been rejected or an error occurred
   * Returned value is saved and can be returned in a promise through promise()
   */
  error(callback: (resp: ErrorObject) => any): ApiHelper {
    this.onError = callback;
    return this;
  }

  /**
   * Call this callback if API promise has been rejected or an error occurred
   * If the callback returns a value it'll overwrite a value returned from
   * success or error callback.
   */
  complete(callback: () => any): ApiHelper {
    this.onComplete = callback;
    return this;
  }

  /**
   * If success, error or complete handler returns a value, promise will be
   * resolved to it.
   * Otherwise boolean will be returned if no errors.
   * This promise will be always resolved even if error occurred.
   * @return {Promise}
   */
  promise() {
    return this.fetchPromise;
  }

  /**
   * Execute a handler inside runInAction and save returned result for promise()
   * @private
   */
  execHandler(callback: Function) {
    runInAction(() => {
      const result = callback();
      if (result !== undefined) {
        this.returnValue = result;
      }
    });
  }

  /**
   * Execute API promise. Here is main helper logic.
   * @protected
   */
  async fetch() {
    if (this.store) {
      if (this.store.error !== undefined) {
        this.store.error = '';
      }
      if (this.store.isLoading !== undefined) {
        this.store.isLoading = true;
      }
    }

    let resp;

    try {
      resp = await this.apiPromise;

      if (resp.ok && !resp.error) {
        this.execHandler(() => this.onSuccess(resp));
      } else {
        //noinspection ExceptionCaughtLocallyJS
        throw new Error(resp.errorDescription);
      }
    } catch (e) {
      const error = resp ? resp.error : 'api_exception';
      const errorDescription = e.message;

      this.execHandler(() => {
        if (this.store && this.store.error !== undefined) {
          this.store.error = errorDescription;
        }
        return this.onError({ error, errorDescription });
      });

      if (!resp) {
        log.warn(`${error} ${errorDescription}`);
      }

      if (this.returnValue === undefined) {
        this.returnValue = false;
      }
    }

    this.execHandler(() => this.onComplete(resp));
    runInAction(() => {
      if (this.store && this.store.isLoading !== undefined) {
        this.store.isLoading = false;
      }
    });

    return this.returnValue !== undefined ? this.returnValue : true;
  }
}


export default function apiHelper(
  apiPromise: Promise<ApiResp>,
  store: Object = null
): ApiHelper {
  const helper = new ApiHelper(apiPromise, store);
  helper.fetchPromise = helper.fetch().catch(log.error);
  return helper;
}

export type ErrorObject = {
  error: string;
  errorDescription: string;
};