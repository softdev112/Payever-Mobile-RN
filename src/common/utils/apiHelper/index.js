import ApiHelper from './ApiHelper';

/**
 * This helper simplifies API call from Store
 * If stateObject is passed this helper will set isLoading and erro fields
 * of this object
 * @param {Function} apiEndPoint
 * @param {Boolean} [showErrorPage]
 * @param {Object} [stateObject]
 * @return {ApiHelper}
 */
export default function apiHelper(
  apiEndPoint: Function,
  stateObject: Object = null,
  showErrorPage: boolean = true
): ApiHelper {
  return new ApiHelper(apiEndPoint, stateObject, showErrorPage);
}