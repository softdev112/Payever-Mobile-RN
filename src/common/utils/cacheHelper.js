import { AsyncStorage } from 'react-native';
import { log } from './index';

let expireData = {};
loadExpireData();

function loadExpireData() {
  AsyncStorage.getItem('cache-expire')
    .then((result) => expireData = JSON.parse(result) || {})
    .catch(() => expireData = {});
}

function saveExpireData(cacheId, lifeTimeSeconds) {
  if (lifeTimeSeconds) {
    const expiredAt = new Date();
    expiredAt.setSeconds(expiredAt.getSeconds() + lifeTimeSeconds);
    expireData[cacheId] = expiredAt;
  } else {
    delete expireData[cacheId];
  }

  AsyncStorage.setItem('cache-expire', JSON.stringify(expireData))
    .catch(log.error);
}

function cacheKey(cacheId) {
  return 'cache:' + cacheId;
}

export default {
  isCacheUpToDate(cacheId, lifeTimeSeconds) {
    const date = expireData ? expireData[cacheId] : null;
    if (!date) {
      return false;
    }

    const secondsPassed = (new Date() - new Date(date)) / 1000;
    return secondsPassed < lifeTimeSeconds;
  },

  isCacheAvailable(cacheId) {
    // TODO: implement this function
    return cacheId;
  },

  loadFromCache(cacheId) {
    return AsyncStorage.getItem(cacheKey(cacheId))
      .then(result => JSON.parse(result))
      .catch(() => null);
  },

  saveToCache(cacheId, lifeTimeSeconds, data) {
    AsyncStorage.setItem(cacheKey(cacheId), JSON.stringify(data))
      .then(() => {
        if (lifeTimeSeconds) {
          saveExpireData(cacheId, lifeTimeSeconds);
        }
      })
      .catch(log.warn);
  },

  removeFromCache(cacheId) {
    AsyncStorage.removeItem(cacheKey(cacheId))
      .then(() => saveExpireData(cacheId))
      .catch(log.warn);
  },
};