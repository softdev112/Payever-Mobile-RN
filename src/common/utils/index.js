/**
 * @providesModule utils
 */

export androidMimeTypes from './androidMimeTypes';
export apiHelper from './apiHelper';
export format from './format';
export log from 'react-native-logging';
export pushNotificationsHelper from './pushNotificationsHelper';
export screenParams from './screenParams';
export soundHelper from './soundHelper';
export deepLinksHelper from './deepLinksHelper';
export networkHelper from './networkHelper';
export cacheHelper from './cacheHelper';