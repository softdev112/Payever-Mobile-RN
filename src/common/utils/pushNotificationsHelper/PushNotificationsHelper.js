import NotificationsIOS, {
  NotificationsAndroid, PendingNotifications,
} from 'react-native-notifications';
import { AppState, Platform } from 'react-native';
import { log } from 'utils';

import type PayeverApi from './../index';
import type UserAccount
  from '../../../store/UserProfilesStore/models/UserAccount';
import handleNotification from './notificationHandlers';

export default class PushNotificationsHelper {
  api: PayeverApi;
  userProfile: UserAccount;

  constructor(api, userProfile) {
    this.api = api;
    this.userProfile = userProfile;
  }

  async onPushNotificationsRegistered(deviceToken) {
    // Send device token to server along side with device name
    log.debug('Device token received:', deviceToken);
    await this.api.device.linkDeviceToken(this.userProfile, deviceToken);
  }

  onNotificationReceivedForeground(notification) {
    log.debug('Receive foreground notification: ', notification);
    handleNotification(notification, { isBackground: false });
  }

  onNotificationReceivedBackground(notification) {
    log.debug('Receive background notification: ', notification);
    handleNotification(notification, { isBackground: true });
  }

  onNotificationOpened(notification) {
    log.debug('On notification opened: ', notification);
    handleNotification(notification, { isOpened: true });
  }

  onNotificationReceivedAndroid(notification) {
    log.debug('Receive Android notification:', AppState.currentState);
    console.log('ddddddddddddddddddd1');
    console.log(notification);
    console.log('ddddddddddddddddddd2');

    if (AppState.currentState === 'active') {
      this.onNotificationReceivedForeground(notification);
    } else {
      this.onNotificationReceivedBackground(notification);
    }
  }

  async registerNotifications() {
    if (Platform.OS === 'ios') {
      NotificationsIOS.addEventListener(
        'remoteNotificationsRegistered',
        ::this.onPushNotificationsRegistered
      );
      NotificationsIOS.requestPermissions();
      NotificationsIOS.consumeBackgroundQueue();

      NotificationsIOS.addEventListener(
        'notificationReceivedForeground',
        ::this.onNotificationReceivedForeground
      );
      NotificationsIOS.addEventListener(
        'notificationReceivedBackground',
        ::this.onNotificationReceivedBackground
      );
      NotificationsIOS.addEventListener(
        'notificationOpened',
        ::this.onNotificationOpened
      );
    } else {
      NotificationsAndroid.setRegistrationTokenUpdateListener(
        ::this.onPushNotificationsRegistered
      );
      NotificationsAndroid.setNotificationOpenedListener(
        ::this.onNotificationOpened
      );
      NotificationsAndroid.setNotificationReceivedListener(
        ::this.onNotificationReceivedAndroid
      );
      const initNotification = await PendingNotifications
        .getInitialNotification();
      console.log('ReactNativeNotif Init Notification', initNotification);
      NotificationsAndroid.refreshToken();
    }
  }
}