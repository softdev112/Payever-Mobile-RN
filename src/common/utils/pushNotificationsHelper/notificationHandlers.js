import { Navigation } from 'react-native-navigation';
import { Platform } from 'react-native';
import { log } from 'utils';

export default function handleNotification(
  notification: NotifIOS | NotifAndroid,
  options: Options
) {
  /* eslint-disable no-underscore-dangle */
  let parameters = null;
  try {
    parameters = Platform.OS === 'ios'
      ? notification._data.parameters
      : JSON.parse(notification.data.parameters);
  } catch (error) {
    log.error(error);
    return;
  }

  if (options.isBackground) {
    // processNotificationData(parameters);
    return;
  }

  // Foreground show notification to user with
  // question does he want to see it
  const message = Platform.OS === 'ios'
    ? notification._alert : 'Receive Notification';

  Navigation.showInAppNotification({
    screen: 'core.PushNotification',
    title: message,
    passProps: {
      message,
      action: () => processNotificationData(parameters),
    },
    navigatorStyle: {
      screenBackgroundColor: 'rgba(0,0,0,0.0)',
    },
    navigatorButtons: {},
    animationType: 'none',
  });

  /* eslint-enable no-underscore-dangle */
}

function processNotificationData(params: NotificationParams) {
  const { subtype, data } =  params;

  if (subtype === 'offer') {
    openOffer(data.offer);
  } else {
    log.warn('Unrecognized notification type');
  }
}

function openOffer(id) {
  // Go to offer preview
  Navigation.showModal({
    screen: 'marketing.ViewOffer',
    title: 'Got an Offer:',
    passProps: {
      offerId: id,
    },
    navigatorStyle: {},
    navigatorButtons: {},
    animationType: 'slide-up',
  });
}

type Options = {
  isBackground: boolean;
};

type NotificationParams = {
  parameters: {
    type: string;
    subtype: string;
    data: Object;
  };
};

type NotifIOS = {
  _alert: string;
  _sound: string;
  _type: string;
  _badge: number;
  _data: NotificationParams;
};

type NotifAndroid = {
  data: {
    collapse_key: string;
    'google.message_id': string;
    'google.sent_time': string;
    parameters: string;
    notification: Object;
  };
};