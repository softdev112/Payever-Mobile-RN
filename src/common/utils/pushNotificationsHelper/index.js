import type PayeverApi from './../index';
import type UserAccount
  from '../../../store/UserProfilesStore/models/UserAccount';
import PushNotificationsHelper from './PushNotificationsHelper';
import { log } from '../index';

export default {
  createInstance,
  getInstance,
};

let instance: PushNotificationsHelper = null;

function createInstance(
  api: PayeverApi,
  userProfile: UserAccount
): PushNotificationsHelper {
  instance = new PushNotificationsHelper(api, userProfile);

  return instance;
}

function getInstance(): PushNotificationsHelper {
  if (!instance) {
    log.error('Push notifications helper error while getting instance');
    throw new Error('Push notifications helper instance was not created');
  }

  return instance;
}