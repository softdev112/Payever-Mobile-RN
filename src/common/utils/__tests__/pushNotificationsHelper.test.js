/* eslint-disable max-len, global-require */
import { Navigation } from 'react-native-navigation';
import { pushNotificationsHelper } from 'utils';
import { AppState, Platform } from 'react-native';
import NotificationsIOS, {
  NotificationsAndroid, PendingNotifications,
} from 'react-native-notifications';
import PushNotificationsHelper
  from '../pushNotificationsHelper/PushNotificationsHelper';
import * as handlers from '../pushNotificationsHelper/notificationHandlers';
import Store from '../../../store';
import config from '../../../config';
import { profilesList } from '../../../store/profiles/__tests__/data';
import UserAccount from '../../../store/profiles/models/UserAccount';

jest.mock('../../../store/auth')
  .mock('AppState', () => ({}))
  .mock('Platform', () => ({
    select: jest.fn(() => {}),
    OS: 'ios',
  }))
  .mock('react-native-device-info', () => ({
    getUniqueID: jest.fn(() => 11111),
    getBrand: jest.fn(() => 'Brand'),
    getModel: jest.fn(() => 'Phone'),
  }))
  .mock('react-native-notifications', () => ({
    addEventListener: jest.fn(),
    requestPermissions: jest.fn(),
    consumeBackgroundQueue: jest.fn(),
    NotificationsAndroid: {
      setRegistrationTokenUpdateListener: jest.fn(),
      setNotificationOpenedListener: jest.fn(),
      setNotificationReceivedListener: jest.fn(),
      refreshToken: jest.fn(),
    },
    PendingNotifications: {
      getInitialNotification: jest.fn(() => ({ data: 'data' })),
    },
  }));

const notificationIOS = {
  _alert: 'Hello world!!!!',
  _sound: 'Sound',
  _type: 'offer',
  _badge: 0,
  _data: {
    parameters: {
      type: 'offer',
      subtype: 'offer',
      data: {
        offer: 11111,
      },
    },
  },
};

describe('Utils/pushNotificationsHelper', () => {
  let store;
  let user;

  beforeAll(() => {
    store = new Store(config);
    user = new UserAccount(profilesList.private.user);
    Navigation.showModal = jest.fn();
    Navigation.showInAppNotification = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('pushNotificationsHelper/getInstance()', () => {
    // It should be the first test before instance of pushNotificationsHelper
    // will be created
    it('getInstance should throw error if instance was NOT created before', () => {
      expect(() => pushNotificationsHelper.getInstance())
        .toThrowError('Push notifications helper was not created');
    });

    it('getInstance should return instance of PushNotificationHelper class if it was created before', () => {
      pushNotificationsHelper.createInstance(store.api, user);
      const pushInstance = pushNotificationsHelper.getInstance();

      expect(pushInstance).not.toBeNull();
      expect(pushInstance).toBeInstanceOf(PushNotificationsHelper);
    });
  });

  describe('pushNotificationsHelper/createInstance(api, userProfile)', () => {
    it('createInstance should create and return an instance of PushNotificationHelper class if arguments != null', () => {
      const pushsHelper = pushNotificationsHelper.createInstance(store.api, user);
      expect(pushsHelper).toBeInstanceOf(PushNotificationsHelper);
    });

    it('createInstance should return null if arguments === null', () => {
      let pushsHelper = pushNotificationsHelper.createInstance(null, user);
      expect(pushsHelper).toBeNull();

      pushsHelper = pushNotificationsHelper.createInstance(store.api, null);
      expect(pushsHelper).toBeNull();
    });
  });

  describe('PushNotificationsHelper/class functions', () => {
    it('onPushNotificationsRegistered(deviceToken) should call proper device api endpoint', async () => {
      const { api } = store;
      const apiSpy = jest.spyOn(api.device, 'linkDeviceToken');
      const postSpy = jest.spyOn(api, 'post');
      api.fetch = jest.fn();

      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      await PushNotificationsHelper.prototype
        .onPushNotificationsRegistered.call(pushHelper, 'Token');

      expect(apiSpy).toHaveBeenCalled();
      expect(apiSpy).toHaveBeenCalledWith(expect.any(Object), 'Token');
      expect(postSpy).toHaveBeenCalled();
      expect(postSpy).toHaveBeenCalledWith(
        '/device/link',
        {
          email: 'payevertest1@gmail.com',
          emailEnabled: false,
          label: 'Brand-Phone',
          phone: '+12345678900',
          platform: expect.any(String),
          pushEnabled: true,
          smsEnabled: false,
          token: 'Token',
          udid: 11111,
        },
        {
          headers: {
            token: undefined,
          },
        }
      );
      expect(api.fetch).toHaveBeenCalled();
      expect(api.fetch).toHaveBeenCalledWith(
        '/device/link',
        expect.any(Object)
      );
    });

    it('onNotificationReceivedForeground(notification) should call handleNotification(notification, { isBackground: false })', () => {
      const handleNotificationSpy = jest.spyOn(handlers, 'default');

      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      PushNotificationsHelper.prototype
        .onNotificationReceivedForeground.call(pushHelper, notificationIOS);

      expect(handleNotificationSpy).toHaveBeenCalled();
      expect(handleNotificationSpy).toHaveBeenCalledWith(
        notificationIOS,
        { isBackground: false }
      );
      expect(Navigation.showInAppNotification).toHaveBeenCalled();
      expect(Navigation.showInAppNotification).toHaveBeenCalledWith(
        {
          animationType: 'none',
          navigatorButtons: expect.any(Object),
          navigatorStyle: expect.any(Object),
          passProps: {
            action: expect.any(Function),
            message: 'Hello world!!!!',
          },
          screen: 'core.PushNotification',
          title: 'Hello world!!!!',
        }
      );
    });

    it('onNotificationReceivedBackground(notification) should call handleNotification(notification, { isBackground: true })', () => {
      const handleNotificationSpy = jest.spyOn(handlers, 'default');

      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      PushNotificationsHelper.prototype
        .onNotificationReceivedBackground.call(pushHelper, notificationIOS);

      expect(handleNotificationSpy).toHaveBeenCalled();
      expect(handleNotificationSpy).toHaveBeenCalledWith(
        notificationIOS,
        { isBackground: true }
      );
      expect(Navigation.showInAppNotification).toHaveBeenCalled();
      expect(Navigation.showInAppNotification).toHaveBeenCalledWith(
        {
          animationType: 'none',
          navigatorButtons: expect.any(Object),
          navigatorStyle: expect.any(Object),
          passProps: {
            action: expect.any(Function),
            message: 'Hello world!!!!',
          },
          screen: 'core.PushNotification',
          title: 'Hello world!!!!',
        }
      );
    });

    it('onNotificationOpened(notification) should call handleNotification(notification, { isOpened: true }) and call Navigation.showModal', () => {
      const handleNotificationSpy = jest.spyOn(handlers, 'default');

      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      PushNotificationsHelper.prototype
        .onNotificationOpened.call(pushHelper, notificationIOS);

      expect(handleNotificationSpy).toHaveBeenCalled();
      expect(handleNotificationSpy).toHaveBeenCalledWith(
        notificationIOS,
        { isOpened: true }
      );
      expect(Navigation.showInAppNotification).not.toHaveBeenCalled();
      expect(Navigation.showModal).toHaveBeenCalled();
      expect(Navigation.showModal).toHaveBeenCalledWith(
        {
          animationType: 'slide-up',
          navigatorButtons: expect.any(Object),
          navigatorStyle: expect.any(Object),
          passProps: {
            offerId: 11111,
          },
          screen: 'marketing.ViewOffer',
          title: 'Got an Offer:',
        }
      );
    });

    it('onNotificationReceivedAndroid(notification) should call onNotificationReceivedForeground if AppState.currentState === active', () => {
      const receivedForegroundSpy = jest.spyOn(
        PushNotificationsHelper.prototype, 'onNotificationReceivedForeground'
      );
      const receivedBackgroundSpy = jest.spyOn(
        PushNotificationsHelper.prototype, 'onNotificationReceivedBackground'
      );

      AppState.currentState = 'active';
      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      PushNotificationsHelper.prototype
        .onNotificationReceivedAndroid.call(pushHelper, notificationIOS);

      expect(receivedForegroundSpy).toHaveBeenCalled();
      expect(receivedBackgroundSpy).not.toHaveBeenCalled();
    });

    it('onNotificationReceivedAndroid(notification) should call onNotificationReceivedBackground if AppState.currentState !== active', () => {
      const receivedForegroundSpy = jest.spyOn(
        PushNotificationsHelper.prototype, 'onNotificationReceivedForeground'
      );
      const receivedBackgroundSpy = jest.spyOn(
        PushNotificationsHelper.prototype, 'onNotificationReceivedBackground'
      );

      AppState.currentState = 'background';
      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      PushNotificationsHelper.prototype
        .onNotificationReceivedAndroid.call(pushHelper, notificationIOS);

      expect(receivedForegroundSpy).not.toHaveBeenCalled();
      expect(receivedBackgroundSpy).toHaveBeenCalled();
    });

    it('registerNotifications() should add right listeners for iOS', () => {
      Platform.OS = 'ios';
      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      PushNotificationsHelper.prototype
        .registerNotifications.call(pushHelper);

      expect(NotificationsIOS.addEventListener).toHaveBeenCalledTimes(4);

      expect(NotificationsIOS.addEventListener.mock.calls[0])
        .toEqual(['remoteNotificationsRegistered', expect.any(Function)]);

      expect(NotificationsIOS.addEventListener.mock.calls[1])
        .toEqual(['notificationReceivedForeground', expect.any(Function)]);

      expect(NotificationsIOS.addEventListener.mock.calls[2])
        .toEqual(['notificationReceivedBackground', expect.any(Function)]);

      expect(NotificationsIOS.addEventListener.mock.calls[3])
        .toEqual(['notificationOpened', expect.any(Function)]);

      expect(NotificationsIOS.requestPermissions).toHaveBeenCalled();
      expect(NotificationsIOS.consumeBackgroundQueue).toHaveBeenCalled();
    });

    it('registerNotifications() should add right listeners for Android', async () => {
      const onNotifOpenSpy = jest.spyOn(
        PushNotificationsHelper.prototype,
        'onNotificationOpened'
      );

      Platform.OS = 'android';
      const pushHelper = pushNotificationsHelper
        .createInstance(store.api, user);
      await PushNotificationsHelper.prototype
        .registerNotifications.call(pushHelper);

      expect(NotificationsAndroid.setRegistrationTokenUpdateListener)
        .toHaveBeenCalled();
      expect(NotificationsAndroid.setRegistrationTokenUpdateListener)
        .toHaveBeenCalledWith(expect.any(Function));

      expect(NotificationsAndroid.setNotificationOpenedListener)
        .toHaveBeenCalled();
      expect(NotificationsAndroid.setNotificationOpenedListener)
        .toHaveBeenCalledWith(expect.any(Function));

      expect(NotificationsAndroid.setNotificationReceivedListener)
        .toHaveBeenCalled();
      expect(NotificationsAndroid.setNotificationReceivedListener)
        .toHaveBeenCalledWith(expect.any(Function));

      expect(NotificationsAndroid.refreshToken).toHaveBeenCalled();
      expect(PendingNotifications.getInitialNotification).toHaveBeenCalled();
      expect(onNotifOpenSpy).toHaveBeenCalled();
      expect(onNotifOpenSpy).toHaveBeenCalledWith({ data: 'data' });
    });
  });
});