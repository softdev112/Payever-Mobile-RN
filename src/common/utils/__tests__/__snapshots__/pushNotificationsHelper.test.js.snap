// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils/pushNotificationsHelper PushNotificationsHelper/class functions registerNotifications() should add right listeners for iOS 1`] = `
Array [
  Array [
    "remoteNotificationsRegistered",
    [Function],
  ],
  Array [
    "notificationReceivedForeground",
    [Function],
  ],
  Array [
    "notificationReceivedBackground",
    [Function],
  ],
  Array [
    "notificationOpened",
    [Function],
  ],
]
`;
